pkg/autoscaler/autoscaler/util_test.go

```diff
@@ -110,87 +110,313 @@ func TestCheckStsAutoScalingInterval(t *testing.T) {
+				sourceTAC := newTidbClusterAutoScaler()
+				sourceTAC.Spec.TiKV.Resources = map[string]v1alpha1.AutoResource{
+					"storage": {
+						CPU:     resource.MustParse("1"),
+						Memory:  resource.MustParse("1G"),
+						Storage: resource.MustParse("10G"),
+					},
+					"no-storage": {
+						CPU:    resource.MustParse("1"),
+						Memory: resource.MustParse("1G"),
+					},
+				}
+				sourceTAC.Spec.TiKV.Rules = map[corev1.ResourceName]v1alpha1.AutoRule{
+					corev1.ResourceCPU: {
+					},
+				}
+				sourceTAC.Spec.TiDB.Resources = map[string]v1alpha1.AutoResource{
+					"no-storage": {
+						CPU:    resource.MustParse("1"),
+						Memory: resource.MustParse("1G"),
+					},
+				}
+				sourceTAC.Spec.TiDB.Rules = map[corev1.ResourceName]v1alpha1.AutoRule{
+					corev1.ResourceCPU: {
+					},
+				}
+				return sourceTAC
+			}(),
+			expectTac: func() *v1alpha1.TidbClusterAutoScaler {
+				expectTac := newTidbClusterAutoScaler()
+				expectTac.Spec.TiKV.Resources = map[string]v1alpha1.AutoResource{
+					"storage": {
+						CPU:     resource.MustParse("1"),
+						Memory:  resource.MustParse("1G"),
+						Storage: resource.MustParse("10G"),
+					},
+					"no-storage": {
+						CPU:    resource.MustParse("1"),
+						Memory: resource.MustParse("1G"),
+					},
+				}
+				expectTac.Spec.TiKV.Rules = map[corev1.ResourceName]v1alpha1.AutoRule{
+					corev1.ResourceCPU: {
+						MinThreshold:  pointer.Float64Ptr(0.1),
+						ResourceTypes: []string{"storage"},
+					},
+				}
+				expectTac.Spec.TiDB.Resources = map[string]v1alpha1.AutoResource{
+					"no-storage": {
+						CPU:    resource.MustParse("1"),
+						Memory: resource.MustParse("1G"),
+					},
+				}
+				expectTac.Spec.TiDB.Rules = map[corev1.ResourceName]v1alpha1.AutoRule{
+					corev1.ResourceCPU: {
+						MinThreshold:  pointer.Float64Ptr(0.1),
+						ResourceTypes: []string{"no-storage"},
+					},
+				}
+				return expectTac
+			}(),
 		},
 	}
-	defaultTAC(tac, tc)
-	g.Expect(tac.Spec.TiKV.Rules[corev1.ResourceStorage].ResourceTypes).Should(ConsistOf([]string{"default_tikv"}))
-	g.Expect(tac.Spec.TiDB.Rules[corev1.ResourceCPU].ResourceTypes).Should(ConsistOf([]string{"default_tidb"}))
+	for _, testcase := range testcases {
+		t.Log(testcase.name)
+		tc := newTidbCluster()
+		defaultTAC(testcase.sourceTac, tc)
+		if testcase.sourceTac.Spec.TiKV == nil {
+			g.Expect(testcase.expectTac.Spec.TiKV).Should(BeNil())
+		} else {
+			g.Expect(testcase.sourceTac.Spec.TiKV.Resources).Should(Equal(testcase.expectTac.Spec.TiKV.Resources))
+			g.Expect(testcase.sourceTac.Spec.TiKV.Rules).Should(Equal(testcase.expectTac.Spec.TiKV.Rules))
+		}
+		if testcase.sourceTac.Spec.TiDB == nil {
+			g.Expect(testcase.expectTac.Spec.TiDB).Should(BeNil())
+		} else {
+			g.Expect(testcase.sourceTac.Spec.TiDB.Resources).Should(Equal(testcase.expectTac.Spec.TiDB.Resources))
+			g.Expect(testcase.sourceTac.Spec.TiDB.Rules).Should(Equal(testcase.expectTac.Spec.TiDB.Rules))
+		}
```

Comment:
I think we should also make sure that the spec is valid by calling `validateTAC`

Reply:
updated.